common/dynamicconfig/dynamicconfigfx/fx.go

```diff
@@ -0,0 +1,108 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package dynamicconfigfx
+
+import (
+	"context"
+	"path/filepath"
+
+	"go.uber.org/fx"
+
+	"github.com/uber/cadence/common/config"
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/dynamicconfig/configstore"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/persistence"
+)
+
+// Module provides fx options for dynamic config initialization
+var Module = fx.Options(fx.Provide(New))
+
+// Params required to build a new dynamic config.
+type Params struct {
+	fx.In
+
+	Cfg     config.Config
+	Logger  log.Logger
+	RootDir string `name:"root-dir"`
+
+	Lifecycle fx.Lifecycle
+}
+
+// New creates dynamicconfig.Client from the configuration
+func New(p Params) dynamicconfig.Client {
+	stopped := make(chan struct{})
+
+	if p.Cfg.DynamicConfig.Client == "" {
+		p.Cfg.DynamicConfigClient.Filepath = constructPathIfNeed(p.RootDir, p.Cfg.DynamicConfigClient.Filepath)
+	} else {
+		p.Cfg.DynamicConfig.FileBased.Filepath = constructPathIfNeed(p.RootDir, p.Cfg.DynamicConfig.FileBased.Filepath)
+	}
+
+	p.Lifecycle.Append(fx.Hook{OnStop: func(_ context.Context) error {
+		close(stopped)
+		return nil
+	}})
+
+	var res dynamicconfig.Client
+
+	var err error
+	if p.Cfg.DynamicConfig.Client == "" {
+		p.Logger.Warn("falling back to legacy file based dynamicClientConfig")
+		res, err = dynamicconfig.NewFileBasedClient(&p.Cfg.DynamicConfigClient, p.Logger, stopped)
+	} else {
+		switch p.Cfg.DynamicConfig.Client {
+		case dynamicconfig.ConfigStoreClient:
+			p.Logger.Info("initialising ConfigStore dynamic config client")
+			res, err = configstore.NewConfigStoreClient(
+				&p.Cfg.DynamicConfig.ConfigStore,
+				&p.Cfg.Persistence,
+				p.Logger,
+				persistence.DynamicConfig,
+			)
+		case dynamicconfig.FileBasedClient:
+			p.Logger.Info("initialising File Based dynamic config client")
+			res, err = dynamicconfig.NewFileBasedClient(&p.Cfg.DynamicConfig.FileBased, p.Logger, stopped)
+		}
+	}
+
+	if res == nil {
+		p.Logger.Info("initialising NOP dynamic config client")
+		res = dynamicconfig.NewNopClient()
+	} else if err != nil {
+		p.Logger.Error("creating dynamic config client failed, using no-op config client instead", tag.Error(err))
```

Comment:
Error cases silently switching to no-op config doesn't sound right. I know that's also the current behavior but we should consider failing startup if there's a misconfiguration for dynamicconfigs.

Reply:
That's a good point, but I'm unsure how safe it is. It's something to consider for v2: a stricter startup to ensure no missing dependencies.